### PR TITLE
Fix/refactor date time

### PR DESCRIPTION
**Pull request by riban-bw**
In response to issue #1545 this PR refactors DateTime to rationalise method names, deprecating old (inconsistent) names.
Moved from Services to SmingCore.
Also adds format method which returns a string based on formatting, providing strftime formatting of string. (Some elements not implemented but mostly there.)
Provides a sample project demonstrating DateTime string output features.
Fixes bug with toHttpDate not setting day of week.
Changed convenience string methods to use new format method.
Not (yet) implemented any locale - all currently in English with existing variation in date formats, e.g. US date format with dot separator used by some European countries.

### CODE REVIEW

**Comment by slaff** [comment]
@riban-bw Don't forget to run `make cs` and commit the coding style changes.

File: Sming/SmingCore/DateTime/DateTime.h [hunk 1]
```diff
@@ -112,14 +112,26 @@ class DateTime
...
 	 */
-	bool parseHttpDate(const String& httpDate);
+	bool fromHttpDate(const String& httpDate);
+
+	/** @brief  Parse a HTTP full date and set time and date
+	 *  @param  httpDate HTTP full date in RFC 1123 format, e.g. Sun, 06 Nov 1994 08:49:37 GMT
+	 *  @retval bool True on success
+	 *  @note   Also supports obsolete RFC 850 date format, e.g. Sunday, 06-Nov-94 08:49:37 GMT where 2 digit year represents range 1970-2069
+	 *  @note   GMT suffix is optional and is always assumed / ignored
+	 *  @deprecated Use 'fromHttpDat' instead
```
Comment (slaff) [inline Sming/SmingCore/DateTime/DateTime.h]: `fromHttpDat` change to `fromHttpDate`.

File: Sming/SmingCore/DateTime/DateTime.cpp [hunk 2]
```diff
@@ -0,0 +1,416 @@
+/*
```
Comment (slaff) [inline Sming/SmingCore/DateTime/DateTime.cpp]: Move `Sming/SmingCore/DateTime/DateTime.*` to `Sming/SmingCore/DateTime.*`. We don't need additional sub-directory for just one class.

File: Sming/SmingCore/SmingCore.h [hunk 3]
```diff
@@ -55,7 +55,8 @@
 #include "Data/Stream/FileStream.h"
```
Comment (slaff) [inline Sming/SmingCore/SmingCore.h]: You not' only change the content of the file but also its permissions. `Sming/SmingCore/SmingCore.h 100644 → 100755`. The file permission change has to be reverted.

File: Sming/SmingCore/DateTime/DateTime.h (hunk 1 above)
Comment (riban-bw) [reply Sming/SmingCore/DateTime/DateTime.h]: Fixed in working copy. Will push when coding style checked and committed.

File: Sming/SmingCore/DateTime/DateTime.cpp (hunk 2 above)
Comment (riban-bw) [reply Sming/SmingCore/DateTime/DateTime.cpp]: Fixed in working copy. Will push when coding style checked and committed. I am currently unable to remove the directory with git. Will investigate further.

File: Sming/SmingCore/SmingCore.h (hunk 3 above)
Comment (riban-bw) [reply Sming/SmingCore/SmingCore.h]: Not sure what happened there. Cygwin seems to do odd things at times. Fixed in working copy. Will push when coding style checked and committed.

File: Sming/SmingCore/DateTime/DateTime.cpp (hunk 2 above)
Comment (slaff) [reply Sming/SmingCore/DateTime/DateTime.cpp]: Try these commands:
```
git mv Sming/SmingCore/DateTime/DateTime.h Sming/SmingCore/DateTime.h
git mv Sming/SmingCore/DateTime/DateTime.cpp Sming/SmingCore/DateTime.cpp
git rm -r Sming/SmingCore/DateTime/
```

**Comment by riban-bw** [comment]
> @riban-bw Don't forget to run `make cs` and commit the coding style changes.

I have downgraded Doxygen and Clang in Cygwin to one that works so I can now run clang-format. Hopefully this should improve my ability to submit changes within Sming coding style guidelines. I may even be able to beat my IDE into submission to produce the required style to begin with!

File: Sming/SmingCore/DateTime/DateTime.h (hunk 1 above)
Comment (riban-bw) [reply Sming/SmingCore/DateTime/DateTime.h]: Fixed in last push.

File: Sming/SmingCore/DateTime/DateTime.cpp (hunk 2 above)
Comment (riban-bw) [reply Sming/SmingCore/DateTime/DateTime.cpp]: I tried rm -r but git doesn't think the directory exists so maybe something I did removed it. It looks like it is now deleted.

File: Sming/SmingCore/SmingCore.h (hunk 3 above)
Comment (riban-bw) [reply Sming/SmingCore/SmingCore.h]: Fixed in last push.

File: Sming/SmingCore/Network/Http/HttpConnection.cpp [hunk 4]
```diff
@@ -89,7 +89,7 @@ DateTime HttpConnection::getLastModifiedDate()
 {
```
Comment (slaff) [inline Sming/SmingCore/Network/Http/HttpConnection.cpp]: `Sming/SmingCore/Network/Http/HttpConnection.cpp 100644 → 100755`

Fix the permissions here.

File: Sming/SmingCore/Network/Http/HttpConnectionBase.cpp [hunk 5]
```diff
@@ -44,7 +44,7 @@ void HttpConnectionBase::init(http_parser_type type)
 #endif
 		parserSettings.on_url = staticOnPath;
 
-		// Data callbacks: on_url, (common) on_header_field, on_header_value, on_body;
+// Data callbacks: on_url, (common) on_header_field, on_header_value, on_body;
```
Comment (slaff) [inline Sming/SmingCore/Network/Http/HttpConnectionBase.cpp]: That change is not needed. Take it out of the PR.

File: Sming/SmingCore/Network/MqttClient.h [hunk 6]
```diff
@@ -309,7 +309,7 @@ class MqttClient : protected TcpClient
 
 	// client flags
 	uint8_t flags = 0;
-	/* 7 8 6 5 4 3 2 1 0
+/* 7 8 6 5 4 3 2 1 0
```
Comment (slaff) [inline Sming/SmingCore/Network/MqttClient.h]: That change is not needed. Take it out of the PR.

File: Sming/SmingCore/Network/NtpClient.h [hunk 7]
```diff
@@ -18,7 +18,7 @@
 #include "UdpConnection.h"
 #include "Platform/System.h"
 #include "Timer.h"
-#include "../Services/DateTime/DateTime.h"
+#include <DateTime.h>
```
Comment (slaff) [inline Sming/SmingCore/Network/NtpClient.h]: Use `#include "DateTime.h"` instead of `#include <DateTime.h>`.

File: Sming/SmingCore/SPI.cpp [hunk 8]
```diff
@@ -119,8 +119,9 @@ uint32 SPIClass::transfer32(uint32 data, uint8 bits)
 	regvalue |= SPI_USR_MOSI | SPI_DOUTDIN | SPI_CK_I_EDGE;
 	WRITE_PERI_REG(SPI_USER(SPI_NO), regvalue);
 
-	WRITE_PERI_REG(SPI_USER1(SPI_NO), (((bits - 1) & SPI_USR_MOSI_BITLEN) << SPI_USR_MOSI_BITLEN_S) |
-										  (((bits - 1) & SPI_USR_MISO_BITLEN) << SPI_USR_MISO_BITLEN_S));
+	WRITE_PERI_REG(SPI_USER1(SPI_NO),
```
Comment (slaff) [inline Sming/SmingCore/SPI.cpp]: Those changes is not needed. `Sming/SmingCore/SPI.cpp` should not be changed in this PR.

File: Sming/SmingCore/SystemClock.h [hunk 9]
```diff
@@ -13,7 +13,7 @@
 #ifndef SMINGCORE_SYSTEMCLOCK_H_
 #define SMINGCORE_SYSTEMCLOCK_H_
 
-#include "../Services/DateTime/DateTime.h"
+#include <DateTime.h>
```
Comment (slaff) [inline Sming/SmingCore/SystemClock.h]: Use #include "DateTime.h" instead of #include <DateTime.h>.

File: Sming/SmingCore/Wire.cpp [hunk 10]
```diff
@@ -271,7 +271,7 @@ void TwoWire::onRequest(void (*function)(void))
 	//user_onRequest = function;
 }
 
-	// Preinstantiate Objects //////////////////////////////////////////////////////
+// Preinstantiate Objects //////////////////////////////////////////////////////
```
Comment (slaff) [inline Sming/SmingCore/Wire.cpp]: Wire.cpp should not be modified in this PR.

File: samples/Basic_DateTime/.project [hunk 11]
```diff
@@ -0,0 +1,28 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<projectDescription>
+	<name>Basic_Blink</name>
```
Comment (slaff) [inline samples/Basic_DateTime/.project]: `<name>Basic_Blink</name>`: Change the name to Basic_DateTime.

File: samples/Basic_DateTime/app/application.cpp [hunk 12]
```diff
@@ -0,0 +1,119 @@
+/**	Basic_DateTime sample application
+	Author: Brian Walton (brian@riban.co.uk)
+	Date: 2018-12-10
+	Provides serial interface, accepting Unix timestamp
+	Prints each type of DateTime::format option
+*/
+#include <user_config.h>
+#include <SmingCore/SmingCore.h>
+
+time_t g_timestamp = 0;
```
Comment (slaff) [inline samples/Basic_DateTime/app/application.cpp]: Use camelCase. For this variable and the one below.

File: samples/Basic_Delegates/app/application.cpp [hunk 13]
```diff
@@ -53,8 +53,8 @@ class Task
 		taskTimer
```
Comment (slaff) [inline samples/Basic_Delegates/app/application.cpp]: Don't modify this file in the current PR.

File: samples/Basic_ProgMem/app/TestProgmem.cpp [hunk 14]
```diff
@@ -123,7 +123,7 @@ void testFSTR(Print& out)
 	}
 	out.println("}");
 
-	// Test equality operators
+// Test equality operators
```
Comment (slaff) [inline samples/Basic_ProgMem/app/TestProgmem.cpp]: Don't modify this file in the current PR.

File: samples/SDCard/app/application.cpp [hunk 15]
```diff
@@ -106,7 +106,7 @@ void stat_file(char* fname)
 }
 
 FRESULT ls(const char* path /* Start node to be scanned (also used as work area) */
-)
+		   )
```
Comment (slaff) [inline samples/SDCard/app/application.cpp]: Don't modify this file in the current PR.

**Comment by riban-bw** [comment]
I sometimes hate git - grrrrrr! The hard part should be writing the code not committing it to version control. Hopefully it looks in better shape now.
I have added one of the missing format parameters (%U week number) which leaves two others (both permutations of week number). I also added locale which may be useful in other modules. I have added what looks right to me for USA, UK, Germany, France and Australia which are the countries / languages that jump to mind where contributors mostly reside. Please check if my assumptions are correct (blame Google if they are wrong) and offer any missing locale. (There are obviously lots missing but it feels like unproportional effort to implement everywhere in the world on this first commit!)

File: Sming/SmingCore/Network/Http/HttpConnectionBase.cpp (hunk 5 above)
Comment (riban-bw) [reply Sming/SmingCore/Network/Http/HttpConnectionBase.cpp]: Fixed

File: samples/Basic_DateTime/app/application.cpp (hunk 12 above)
Comment (riban-bw) [reply samples/Basic_DateTime/app/application.cpp]: Fixed

File: samples/Basic_Delegates/app/application.cpp (hunk 13 above)
Comment (riban-bw) [reply samples/Basic_Delegates/app/application.cpp]: Fixed

File: samples/Basic_ProgMem/app/TestProgmem.cpp (hunk 14 above)
Comment (riban-bw) [reply samples/Basic_ProgMem/app/TestProgmem.cpp]: Fixed

File: samples/SDCard/app/application.cpp (hunk 15 above)
Comment (riban-bw) [reply samples/SDCard/app/application.cpp]: Fixed

**Comment by riban-bw** [comment]
> But please DON'T use percentage Something (%S) as format. Use "S". Please, see the format parameters for PHP for example and follow them http://php.net/manual/en/function.date.php#format .

Sming is a C++ framework. I think the strftime format may be familiar to C++ programmers. It is also more flexible than the PHP implementation - normal text may be included within the format string, e.g. "Time: %X". (I prefer Sming to Arduino because Sming implements printf which is an invaluable feature.) I understand that someone who is familiar with PHP may prefer the PHP format, someone who is familiar with Python may want the Python format, etc. but I recommend implementing the % format for all those C++ programmers who are already familiar with it. We could implement a similar function with different formatting if you want (under a separate PR). This could be designed to avoid overheads due to the way that only used functions are added to the firmware.

Regarding locale - this is implemented mostly with those horrid pre-compiler directives which ensure only the required bits are put into the code, e.g. `#define LOCALE LOCALE_DE_DE` should only include the German locale elements where they are actually used by the firmware. I don't think that Locale.h adds any bloat to the library as it simply defines **only** the strings that are already used in DateTime and if a user does not call _format_ in their source code then it won't be added to the firmware.

I agree that we need to constrain firmware size with particular attention to each type of memory and do try to code with an _embedded_ ethos, though may not create optimal code on a first attempt. (That is what peer review is for - right?) On the point of bloat, I was considering adding a feature request to measure libsming size after each commit and compare with previous / develop head, so that we can monitor the change in size that each commit has which may help to keep a handle on bloat.

**Comment by slaff** [comment]
> On the point of bloat, I was considering adding a feature request to measure libsming size after each commit and compare with previous / develop head, so that we can monitor the change in size that each commit has which may help to keep a handle on bloat.

Do you know a free web service where we can send the data and it will generated the needed graphs for us? It is good to see how data size, free heap size and code size vary.

**Comment by slaff** [comment]
> Sming is a C++ framework. I think the strftime format may be familiar to C++ programmers.

I agree with you. Let it be `http://www.cplusplus.com/reference/ctime/strftime/`

One more point - make sure that the code passes CI tests. At the moment it seem to fail under Travis due to coding style issues and under Windows due to 
```c++
'setlocale' is not a member of 'std'
     char* __old = std::setlocale(LC_NUMERIC, 0);
                   ^
```

**Comment by riban-bw** [comment]
I can see that the coding style issue relates to indentation I added to Locale.h to make things easier to understand and edit. I will fix that (and also fix some errors with German I was able to identify when chatting to a German colleague). The Windows issue may be more tricky. I looks like something peculiar to Chocolatey which I cannot install on my development machine (the only Windows machine I have access to). This error does not occur with cygwin or Ubuntu. I suspect that the use of Unicode characters in some of the languages may be triggering espressif libraries to kick in locale elements which fail for Windows. This may be a dormant bug in their software. I will look at the errors from appveyor and try to diagnose but this may prove challenging. If anyone with a Chocolatey install is available to help diagnose and test this it would be useful.

**Comment by slaff** [comment]
> I suspect that the use of Unicode characters in some of the languages may be triggering xtensa libraries to kick in locale elements which fail for Windows

I guess the issue is more trivial. On Windows the files names are case insensitive. Which means that your Locale.h will be loaded when locale.h. Try renaming it to something like CoreLocale.h and see if that works.

**Comment by riban-bw** [comment]
Thanks slaff, I will try that. (I configured my Windows machine to be case insensitive to allow me to build xtensa. Maybe I should change it back to pick up stuff like this.) I was about to commit these changes but realised I had started to implement the last couple of formats (variations on week number) but not yet tested so need to complete that before I commit. Hopefully sort in next few days. (I didn't have to change the German - I had got it right first time! Native German speakers feel free to contradict me.)

**Comment by slaff** [comment]
> German - I had got it right first time

The months and the days of the week are correct. If there is something more I can check it too.

**Comment by riban-bw** [comment]
The API documentation has a table showing which parameters are influenced by locale. I guess I could change the locale of a test machine and compare Sming output with a native test app. (I think this can be done with an environmental variable in Linux so I could probably create a test for this for the supported languages.) I have been doing similar with cygwin but its implementation of strftime seems simplistic and does not include all the options so firing up a Linux machine may be a better test environment. The only other locale element I set for Germany is the short date where Germany uses the dot separator.

**Comment by slaff** [comment]
> samples/SDCard/app/application.cpp 100644 → 100755

That unneeded modification is still in the PR. Remove it using:

```
cd $SMING_HOME
git checkout cc7af2263e41e24c12c047878c95f13489d5c81c
```

After that the older version of the file will be ready to commit to this PR.

**Comment by riban-bw** [comment]
Thanks for the tip @slaff. I used 
```
git diff cc7af2263e41e24c12c047878c95f13489d5c81c|grep -A 1 -B 1 "old mode"
```
to list all the changed permission files and found a few more so corrected those. Knowing the ID of the version to compare to was handy. I have just seen in github where to find that so maybe I can be more efficient with git in future! Life was so much easier with subversion :-).

**Comment by slaff** [comment]
@riban-bw Thanks for the PR. It took a bit longer but hopefully you are now better prepared `git` user :).

**Pull request merged**